scan_rr.go

```diff
@@ -1216,6 +1217,115 @@ func (rr *DS) parse(c *zlexer, o string) *ParseError      { return rr.parseDS(c,
 func (rr *DLV) parse(c *zlexer, o string) *ParseError     { return rr.parseDS(c, o, "DLV") }
 func (rr *CDS) parse(c *zlexer, o string) *ParseError     { return rr.parseDS(c, o, "CDS") }
 
+func (rr *IPSECKEY) parse(c *zlexer, o string) *ParseError {
+	for i := 0; i < 3; i++ {
+		l, _ := c.Next()
+		num, err := strconv.ParseUint(l.token, 10, 8)
+		if err != nil || l.err {
+			return &ParseError{"", "bad IPSECKEY value", l}
+		}
+
+		num8 := uint8(num)
+		switch i {
+		case 0:
+			rr.Precedence = num8
+		case 1:
+			rr.GatewayType = num8
+		case 2:
+			rr.Algorithm = num8
+		}
+
+		c.Next() // zBlank
+	}
+
+	l, _ := c.Next()
+	if l.err {
+		return &ParseError{"", "bad IPSECKEY gateway", l}
+	}
+
+	var err error
+	rr.GatewayAddr, rr.GatewayHost, err = parseAddrHostUnion(l.token, o, rr.GatewayType)
+	if err != nil {
+		return &ParseError{"", "AMTRELAY " + err.Error(), l}
+	}
+
+	c.Next() // zBlank
+
+	s, pErr := endingToString(c, "bad IPSECKEY PublicKey")
+	if pErr != nil {
+		return pErr
+	}
+	rr.PublicKey = s
+	return slurpRemainder(c)
+}
+
+func (rr *AMTRELAY) parse(c *zlexer, o string) *ParseError {
+	for i := 0; i < 3; i++ {
+		l, _ := c.Next()
+		num, err := strconv.ParseUint(l.token, 10, 8)
+		if err != nil || l.err {
+			return &ParseError{"", "bad AMTRELAY value", l}
+		}
+
+		num8 := uint8(num)
+		switch i {
+		case 0:
+			rr.Precedence = num8
+		case 1:
+			if num8 > 1 {
+				return &ParseError{"", "bad discovery value", l}
+			}
+			rr.DiscoveryOptional = num8 == 1
+		case 2:
+			if num > 0x7f {
+				return &ParseError{"", "bad gateway type value", l}
+			}
+			rr.GatewayType = num8
+		}
+
+		c.Next() // zBlank
+	}
+
+	l, _ := c.Next()
+	if l.err {
+		return &ParseError{"", "bad AMTRELAY gateway", l}
+	}
+
+	var err error
+	rr.GatewayAddr, rr.GatewayHost, err = parseAddrHostUnion(l.token, o, rr.GatewayType)
+	if err != nil {
+		return &ParseError{"", "AMTRELAY " + err.Error(), l}
+	}
+
+	return slurpRemainder(c)
+}
+
+// same constants and parsing between IPSECKEY and AMTRELAY
+func parseAddrHostUnion(token, o string, gatewayType uint8) (addr net.IP, host string, err error) {
+	switch gatewayType {
+	case IPSECGatewayNone:
+		if token != "." {
+			return addr, host, errors.New("gateway type none with gateway set")
+		}
+	case IPSECGatewayIPv4, IPSECGatewayIPv6:
+
+		if addr = net.ParseIP(token); addr == nil {
+			return addr, host, errors.New("gateway IP invalid")
+		}
+		if !((gatewayType == IPSECGatewayIPv4 && addr.To4() != nil) || (gatewayType == IPSECGatewayIPv6 && addr.To16() != nil)) {
```

Comment:
`To16` will always return non-nil for the result of `net.ParseIP`. If you're trying to make sure this isn't an IPv4 mapped address, use `To4() == nil`. Otherwise, just drop the `IPSECGatewayIPv6` check.

Reply:
Is the new solution sufficient?